libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java

```diff
@@ -217,8 +223,88 @@ public URL nextElement() {
+        InMemoryModuleFinder moduleFinder = InMemoryModuleFinder.of(missingModules, modulePath);
+        if (modulePath[0].getFileSystem().provider().getScheme().equals("jar")) {
+            modulePath[0].getFileSystem().close();
+        }
+        return moduleFinder;
+    }
+
+    /**
+     * Returns the base prefix for a versioned prefix. Otherwise, the given prefix if not versioned.
+     * For example, given "IMPL-JARS/x-content/jackson-core-2.13.2.jar/META-INF/versions/9", returns
+     * "IMPL-JARS/x-content/jackson-core-2.13.2.jar".
+     */
+    static String basePrefix(String prefix) {
+        int idx = prefix.indexOf(MRJAR_VERSION_PREFIX);
+        if (idx == -1) {
+            return prefix;
+        }
+        return prefix.substring(0, idx - 1);
+    }
+
+    private Path[] modulePath() throws IOException {
+        Function<Path, Path[]> entries = path -> prefixToCodeBase.keySet()
+            .stream()
+            .map(EmbeddedImplClassLoader::basePrefix)
+            .distinct()
+            .map(pfx -> path.resolve(pfx))
+            .toArray(Path[]::new);
+        URI rootURI = rootURI(prefixToCodeBase.values().stream().findFirst().map(CodeSource::getLocation).orElseThrow());
+        if (rootURI.getScheme().equals("file")) {
+            return entries.apply(Path.of(rootURI));
+        } else if (rootURI.getScheme().equals("jar")) {
+            FileSystem fileSystem = FileSystems.newFileSystem(rootURI, Map.of(), ClassLoader.getSystemClassLoader());
+            Path rootPath = fileSystem.getPath("/");
+            return entries.apply(rootPath);
+        } else {
+            throw new IOException("unknown scheme:" + rootURI.getScheme());
+        }
+    }
+
     // -- infra
 
+    /**
+     * Returns the root URI for a given url. The root URI is the base URI where all classes and
+     * resources can be searched for by appending a prefixes.
+     *
+     * Depending on whether running from a jar (distribution), or an exploded archive (testing),
+     * the given url will have one of two schemes, "file", or "jar:file". For example:
+     *  distro- jar:file:/xxx/distro/lib/elasticsearch-x-content-8.2.0-SNAPSHOT.jar!/IMPL-JARS/x-content/xlib-2.10.4.jar
+     *  rootURI jar:file:/xxx/distro/lib/elasticsearch-x-content-8.2.0-SNAPSHOT.jar
+     *
+     *  test  - file:/x/git/es_modules/libs/x-content/build/generated-resources/impl/IMPL-JARS/x-content/xlib-2.10.4.jar
+     *  rootURI file:/x/git/es_modules/libs/x-content/build/generated-resources/impl
+     */
+    static URI rootURI(URL url) {
+        try {
+            URI uri = url.toURI();
+            if (uri.getScheme().equals("jar")) {
+                String s = uri.toString();
+                return URI.create(s.substring(0, s.lastIndexOf("!/")));
+            } else {
+                return URI.create(getParent(getParent(getParent(uri.toString()))));
+            }
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
```

Comment:
Since this is coming from a file, this should not be possible right? It could be an AssertionError?

Reply:
Done.